Add some TZONE_ALLOCATED enforcement.

MenloDorian opened the pull request:
#### 61117dc2de25c1caca02cb1ae82437a45df3f232
<pre>
Add some TZONE_ALLOCATED enforcement.
<a href="https://bugs.webkit.org/show_bug.cgi?id=304684">https://bugs.webkit.org/show_bug.cgi?id=304684</a>
<a href="https://rdar.apple.com/167161285">rdar://167161285</a>

Reviewed by Justin Michaud and Keith Miller.

This is an invariant of TZone allocation as a hardening mechanism: if a class / structure is
TZONE_ALLOCATED, its subclasses should also be TZONE_ALLOCATED.  This patch adds some
verification in MAKE_FAST_ALLOCATED to ensure that the class does not inherit from an ancestor
class that is TZONE_ALLOCATED.  This verification can also be added in other non-TZone
allocation macros, but we&apos;re starting with MAKE_FAST_ALLOCATED for now.

Also fix some classes that should be TZONE_ALLOCATED (found by this verification).
Also fix some incorrect uses of WTF_DEPRECATED_MAKE_FAST_ALLOCATED and
WTF_DEPRECATED_MAKE_STRUCT_FAST_ALLOCATED_WITH_HEAP_IDENTIFIER, where the classname argument
is wrong.  These were copy paste errors that had gone unnoticed until now.  This issue was
found by this verification.

Also introduce WTF_FORBID_HEAP_ALLOCATION_WITH_VALIDATION which is equivalent to
WTF_FORBID_HEAP_ALLOCATION with the !usesTZoneHeap() validation.  Eventually, we will convert
all uses of WTF_FORBID_HEAP_ALLOCATION to add the typename parameter, and replace
WTF_FORBID_HEAP_ALLOCATION_WITH_VALIDATION with WTF_FORBID_HEAP_ALLOCATION.

Also introduce WTF_FORBID_HEAP_ALLOCATION_FOR_ABSTRACT_CLASS which is the equivalent of
WTF_FORBID_HEAP_ALLOCATION_WITH_VALIDATION.  The only reason for this &quot;duplicate&quot; macro is
because it documents the reason why we&apos;re forbidding new and delete operators i.e. the class is
an abstract class.

Also introduce WTF_MAKE_INHERITED_TZONE_ALLOCATED which can be used for subclasses that inherit
from a TZONE_ALLOCATED class but does not add any fields.  Hence. it is identical in shape and
size as its parent.  Therefore, there&apos;s not much added value in declaring its own s_heapRef and
set of TZone support functions.  Just use its parent&apos;s.

No new test.  This patch adds verification that existing tests more effective for catching this
class of issues as well.

* Source/JavaScriptCore/bytecode/ExecutionCounter.h:
* Source/JavaScriptCore/dfg/DFGNode.h:
* Source/JavaScriptCore/dfg/DFGSpeculativeJIT.cpp:
* Source/JavaScriptCore/dfg/DFGSpeculativeJIT.h:
* Source/JavaScriptCore/heap/MarkedBlock.h:
* Source/JavaScriptCore/inspector/augmentable/AlternateDispatchableAgent.h:
* Source/WTF/wtf/FastMalloc.h:
(WTF::usesTZoneHeap):
* Source/WTF/wtf/ForbidHeapAllocation.h:
* Source/WTF/wtf/MetaAllocator.cpp:
* Source/WTF/wtf/MetaAllocatorHandle.h:
* Source/WTF/wtf/SegmentedVector.h:
* Source/WTF/wtf/TZoneMalloc.h:
* Source/WebCore/Modules/WebGPU/Implementation/WebGPUXRProjectionLayerImpl.cpp:
* Source/WebCore/Modules/WebGPU/Implementation/WebGPUXRProjectionLayerImpl.h:
* Source/WebCore/Scripts/SettingsTemplates/Settings.cpp.erb:
* Source/WebCore/Scripts/SettingsTemplates/Settings.h.erb:
* Source/WebCore/css/CSSFontFaceSource.cpp:
* Source/WebCore/css/CSSFontFaceSource.h:
* Source/WebCore/css/CSSSelector.cpp:
* Source/WebCore/css/CSSSelector.h:
* Source/WebCore/fileapi/AsyncFileStream.cpp:
* Source/WebCore/html/HTMLMediaElement.cpp:
* Source/WebCore/loader/FontLoadRequest.h:
* Source/WebCore/loader/LoaderMalloc.cpp:
* Source/WebCore/loader/LocalFrameLoaderClient.h:
* Source/WebCore/page/Page.cpp:
* Source/WebCore/page/PrewarmInformation.h:
* Source/WebCore/page/ProcessWarming.cpp:
* Source/WebCore/platform/graphics/FontCache.cpp:
* Source/WebCore/platform/graphics/LegacyCDMSession.h:
* Source/WebCore/platform/graphics/angle/GraphicsContextGLANGLE.h:
* Source/WebCore/platform/graphics/cocoa/GraphicsContextGLCocoa.h:
* Source/WebCore/platform/graphics/cocoa/GraphicsContextGLCocoa.mm:
* Source/WebCore/platform/graphics/cocoa/WebProcessGraphicsContextGLCocoa.mm:
* Source/WebCore/platform/graphics/gbm/DRMDeviceManager.h:
* Source/WebCore/platform/graphics/gbm/MemoryMappedGPUBuffer.h:
* Source/WebCore/platform/graphics/texmap/BitmapTexturePool.h:
* Source/WebCore/platform/graphics/texmap/GraphicsContextGLTextureMapperANGLE.cpp:
* Source/WebCore/platform/graphics/texmap/GraphicsContextGLTextureMapperANGLE.h:
* Source/WebCore/platform/graphics/texmap/TextureMapperLayer3DRenderingContext.h:
* Source/WebCore/platform/graphics/texmap/coordinated/CoordinatedPlatformLayerBuffer.h:
* Source/WebCore/platform/graphics/win/MediaPlayerPrivateMediaFoundation.cpp:
(WebCore::MediaPlayerPrivateMediaFoundation::AsyncCallback::AsyncCallback):
* Source/WebCore/platform/ios/PlaybackSessionInterfaceTVOS.cpp:
* Source/WebCore/platform/ios/PlaybackSessionInterfaceTVOS.h:
* Source/WebCore/platform/ios/VideoPresentationInterfaceTVOS.h:
* Source/WebCore/platform/ios/VideoPresentationInterfaceTVOS.mm:
* Source/WebCore/rendering/LegacyRootInlineBox.cpp:
* Source/WebCore/style/StyleTreeResolver.h:
* Source/WebCore/xml/XMLHttpRequestProgressEventThrottle.h:
* Source/WebKit/GPUProcess/media/RemoteLegacyCDMSessionProxy.cpp:
* Source/WebKit/GPUProcess/media/RemoteLegacyCDMSessionProxy.h:
* Source/WebKit/NetworkProcess/Authentication/AuthenticationManager.cpp:
* Source/WebKit/UIProcess/Cocoa/PageClientImplCocoa.h:
* Source/WebKit/UIProcess/Cocoa/PageClientImplCocoa.mm:
* Source/WebKit/UIProcess/Inspector/socket/RemoteInspectorProtocolHandler.cpp:
* Source/WebKit/UIProcess/ios/PageClientImplIOS.h:
* Source/WebKit/UIProcess/ios/PageClientImplIOS.mm:
* Source/WebKit/UIProcess/mac/PageClientImplMac.h:
* Source/WebKit/UIProcess/mac/PageClientImplMac.mm:
* Source/WebKit/WebProcess/WebPage/WebHistoryItemClient.cpp:
* Source/WebKit/WebProcess/WebPage/WebHistoryItemClient.h:
* Source/bmalloc/bmalloc/TZoneHeap.h:
(bmalloc::TZone::sizeClassFor):
(bmalloc::TZone::sizeClass):
(bmalloc::TZone::alignment):
(bmalloc::TZone::inheritedSizeClass):
(bmalloc::TZone::inheritedAlignment):
(bmalloc::TZone::usesTZoneHeap):
(bmalloc::api::SizeAndAlignment::encode):
(bmalloc::api::sizeClassFor): Deleted.
* Source/bmalloc/bmalloc/TZoneHeapInlines.h:
* Source/bmalloc/bmalloc/TZoneHeapManager.cpp:
(bmalloc::api::TZoneHeapManager::TZoneHeapManager::heapRefForTZoneTypeDifferentSize):
* Tools/TestWebKitAPI/Tests/WebCore/cocoa/TestGraphicsContextGLCocoa.mm:

Canonical link: <a href="https://commits.webkit.org/304989@main">https://commits.webkit.org/304989@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/b3e77114d1d7963ba61e2f158744030012703243

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows | Apple Internal |
| ----- | ---------------------- | ------- |  ----- |  --------- | ------ |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/137108 "2 style errors") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/159/builds/9468 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/48395 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/144851 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/90081 "Built successfully") | [✅ 🛠 ios-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/d2560935-3116-4e8c-97d4-eae9446ae8d2) 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/138980 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/155/builds/10171 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/156/builds/9595 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/104852 "Passed tests") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/60/builds/76483 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 mac-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/f95b15bb-35c7-4843-94cf-7da5417ba381) 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/140053 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/162/builds/7480 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/122864 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/85688 "Passed tests") | | [✅ 🛠 vision-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/a9ccb474-920f-46c6-98cb-fe9e30bae938) 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/154/builds/7117 "Passed tests") | [✅ 🧪 api-mac-debug](https://ews-build.webkit.org/#/builders/165/builds/4824 "Passed tests") | [✅ 🛠 wpe-cairo-libwebrtc](https://ews-build.webkit.org/#/builders/166/builds/5441 "Built successfully") | | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/129069 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/116466 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/41030 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/147607 "Built successfully") | | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/135595 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/153/builds/9151 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/41595 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/113206 "Passed tests") | | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/160/builds/9169 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/161/builds/7692 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/113536 "Passed tests") | | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/164/builds/7040 "Passed tests") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/119125 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/63503 "Built successfully") | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/21126 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/158/builds/9199 "Built successfully") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/37179 "Passed tests") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/168375 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/157/builds/8923 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/72765 "Built successfully") | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/43947 "Passed tests") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/163/builds/9140 "Built successfully") | | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/152/builds/8992 "Built successfully") | | | | 
<!--EWS-Status-Bubble-End-->

kmiller68 commented on Source/JavaScriptCore/bytecode/ExecutionCounter.h:
```diff
@@ -55,7 +55,7 @@ inline int32_t formattedTotalExecutionCount(float value)
 template<CountingVariant countingVariant>
 class ExecutionCounter {
-    WTF_MAKE_TZONE_ALLOCATED_TEMPLATE(ExecutionCounter);
+    WTF_FORBID_NEW_DELETE_OPERATORS(ExecutionCounter);
```
I think we should name this something like `WTF_FORBID_HEAP_ALLOCATION_WITH_VALIDATION` and add a FIXME to replace `WTF_FORBID_HEAP_ALLOCATION` with this. Alternatively, we could rename `WTF_FORBID_HEAP_ALLOCATION` to `WTF_FORBID_HEAP_ALLOCATION_DEPRECATED` with a comment to switch to this and name this `WTF_FORBID_HEAP_ALLOCATION`

justinmichaud commented on Source/WTF/wtf/FastMalloc.h:
```diff
@@ -354,6 +354,16 @@ struct NonDestructingDeleter {
+template<typename T>
+inline constexpr bool usesTZoneHeap()
+{
+    if constexpr (requires { std::remove_pointer_t<T>::usesTZoneHeap; })
```
I wonder if we could make tests for this. I guess it would be a bit tricky, I don't want to block this patch, but I think things like this should be testable. Perhaps this could be done in combination with the safer cpp tests.

MenloDorian replied:
This code is self testing in its application and tested simply by building WebKit.  That's how I found all the broken cases for FAST_ALLOCATED and inherited TZONE_ALLOCATED.  I don't think a standalone test is needed.

Commit:
Add some TZONE_ALLOCATED enforcement.
https://bugs.webkit.org/show_bug.cgi?id=304684
rdar://167161285

Reviewed by Justin Michaud and Keith Miller.

This is an invariant of TZone allocation as a hardening mechanism: if a class / structure is
TZONE_ALLOCATED, its subclasses should also be TZONE_ALLOCATED.  This patch adds some
verification in MAKE_FAST_ALLOCATED to ensure that the class does not inherit from an ancestor
class that is TZONE_ALLOCATED.  This verification can also be added in other non-TZone
allocation macros, but we're starting with MAKE_FAST_ALLOCATED for now.

Also fix some classes that should be TZONE_ALLOCATED (found by this verification).
Also fix some incorrect uses of WTF_DEPRECATED_MAKE_FAST_ALLOCATED and
WTF_DEPRECATED_MAKE_STRUCT_FAST_ALLOCATED_WITH_HEAP_IDENTIFIER, where the classname argument
is wrong.  These were copy paste errors that had gone unnoticed until now.  This issue was
found by this verification.

Also introduce WTF_FORBID_HEAP_ALLOCATION_WITH_VALIDATION which is equivalent to
WTF_FORBID_HEAP_ALLOCATION with the !usesTZoneHeap() validation.  Eventually, we will convert
all uses of WTF_FORBID_HEAP_ALLOCATION to add the typename parameter, and replace
WTF_FORBID_HEAP_ALLOCATION_WITH_VALIDATION with WTF_FORBID_HEAP_ALLOCATION.

Also introduce WTF_FORBID_HEAP_ALLOCATION_FOR_ABSTRACT_CLASS which is the equivalent of
WTF_FORBID_HEAP_ALLOCATION_WITH_VALIDATION.  The only reason for this "duplicate" macro is
because it documents the reason why we're forbidding new and delete operators i.e. the class is
an abstract class.

Also introduce WTF_MAKE_INHERITED_TZONE_ALLOCATED which can be used for subclasses that inherit
from a TZONE_ALLOCATED class but does not add any fields.  Hence. it is identical in shape and
size as its parent.  Therefore, there's not much added value in declaring its own s_heapRef and
set of TZone support functions.  Just use its parent's.

No new test.  This patch adds verification that existing tests more effective for catching this
class of issues as well.

* Source/JavaScriptCore/bytecode/ExecutionCounter.h:
* Source/JavaScriptCore/dfg/DFGNode.h:
* Source/JavaScriptCore/dfg/DFGSpeculativeJIT.cpp:
* Source/JavaScriptCore/dfg/DFGSpeculativeJIT.h:
* Source/JavaScriptCore/heap/MarkedBlock.h:
* Source/JavaScriptCore/inspector/augmentable/AlternateDispatchableAgent.h:
* Source/WTF/wtf/FastMalloc.h:
(WTF::usesTZoneHeap):
* Source/WTF/wtf/ForbidHeapAllocation.h:
* Source/WTF/wtf/MetaAllocator.cpp:
* Source/WTF/wtf/MetaAllocatorHandle.h:
* Source/WTF/wtf/SegmentedVector.h:
* Source/WTF/wtf/TZoneMalloc.h:
* Source/WebCore/Modules/WebGPU/Implementation/WebGPUXRProjectionLayerImpl.cpp:
* Source/WebCore/Modules/WebGPU/Implementation/WebGPUXRProjectionLayerImpl.h:
* Source/WebCore/Scripts/SettingsTemplates/Settings.cpp.erb:
* Source/WebCore/Scripts/SettingsTemplates/Settings.h.erb:
* Source/WebCore/css/CSSFontFaceSource.cpp:
* Source/WebCore/css/CSSFontFaceSource.h:
* Source/WebCore/css/CSSSelector.cpp:
* Source/WebCore/css/CSSSelector.h:
* Source/WebCore/fileapi/AsyncFileStream.cpp:
* Source/WebCore/html/HTMLMediaElement.cpp:
* Source/WebCore/loader/FontLoadRequest.h:
* Source/WebCore/loader/LoaderMalloc.cpp:
* Source/WebCore/loader/LocalFrameLoaderClient.h:
* Source/WebCore/page/Page.cpp:
* Source/WebCore/page/PrewarmInformation.h:
* Source/WebCore/page/ProcessWarming.cpp:
* Source/WebCore/platform/graphics/FontCache.cpp:
* Source/WebCore/platform/graphics/LegacyCDMSession.h:
* Source/WebCore/platform/graphics/angle/GraphicsContextGLANGLE.h:
* Source/WebCore/platform/graphics/cocoa/GraphicsContextGLCocoa.h:
* Source/WebCore/platform/graphics/cocoa/GraphicsContextGLCocoa.mm:
* Source/WebCore/platform/graphics/cocoa/WebProcessGraphicsContextGLCocoa.mm:
* Source/WebCore/platform/graphics/gbm/DRMDeviceManager.h:
* Source/WebCore/platform/graphics/gbm/MemoryMappedGPUBuffer.h:
* Source/WebCore/platform/graphics/texmap/BitmapTexturePool.h:
* Source/WebCore/platform/graphics/texmap/GraphicsContextGLTextureMapperANGLE.cpp:
* Source/WebCore/platform/graphics/texmap/GraphicsContextGLTextureMapperANGLE.h:
* Source/WebCore/platform/graphics/texmap/TextureMapperLayer3DRenderingContext.h:
* Source/WebCore/platform/graphics/texmap/coordinated/CoordinatedPlatformLayerBuffer.h:
* Source/WebCore/platform/graphics/win/MediaPlayerPrivateMediaFoundation.cpp:
(WebCore::MediaPlayerPrivateMediaFoundation::AsyncCallback::AsyncCallback):
* Source/WebCore/platform/ios/PlaybackSessionInterfaceTVOS.cpp:
* Source/WebCore/platform/ios/PlaybackSessionInterfaceTVOS.h:
* Source/WebCore/platform/ios/VideoPresentationInterfaceTVOS.h:
* Source/WebCore/platform/ios/VideoPresentationInterfaceTVOS.mm:
* Source/WebCore/rendering/LegacyRootInlineBox.cpp:
* Source/WebCore/style/StyleTreeResolver.h:
* Source/WebCore/xml/XMLHttpRequestProgressEventThrottle.h:
* Source/WebKit/GPUProcess/media/RemoteLegacyCDMSessionProxy.cpp:
* Source/WebKit/GPUProcess/media/RemoteLegacyCDMSessionProxy.h:
* Source/WebKit/NetworkProcess/Authentication/AuthenticationManager.cpp:
* Source/WebKit/UIProcess/Cocoa/PageClientImplCocoa.h:
* Source/WebKit/UIProcess/Cocoa/PageClientImplCocoa.mm:
* Source/WebKit/UIProcess/Inspector/socket/RemoteInspectorProtocolHandler.cpp:
* Source/WebKit/UIProcess/ios/PageClientImplIOS.h:
* Source/WebKit/UIProcess/ios/PageClientImplIOS.mm:
* Source/WebKit/UIProcess/mac/PageClientImplMac.h:
* Source/WebKit/UIProcess/mac/PageClientImplMac.mm:
* Source/WebKit/WebProcess/WebPage/WebHistoryItemClient.cpp:
* Source/WebKit/WebProcess/WebPage/WebHistoryItemClient.h:
* Source/bmalloc/bmalloc/TZoneHeap.h:
(bmalloc::TZone::sizeClassFor):
(bmalloc::TZone::sizeClass):
(bmalloc::TZone::alignment):
(bmalloc::TZone::inheritedSizeClass):
(bmalloc::TZone::inheritedAlignment):
(bmalloc::TZone::usesTZoneHeap):
(bmalloc::api::SizeAndAlignment::encode):
(bmalloc::api::sizeClassFor): Deleted.
* Source/bmalloc/bmalloc/TZoneHeapInlines.h:
* Source/bmalloc/bmalloc/TZoneHeapManager.cpp:
(bmalloc::api::TZoneHeapManager::TZoneHeapManager::heapRefForTZoneTypeDifferentSize):
* Tools/TestWebKitAPI/Tests/WebCore/cocoa/TestGraphicsContextGLCocoa.mm:

Canonical link: https://commits.webkit.org/304989@main
